include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,237 @@
+            {
+                const size_t initial_size = ranges::size(rng);
+                const size_t n = static_cast<size_t>(this->n);
+
+                return initial_size > n
+                    ? initial_size - n
+                    : 0;
+            }
+        public:
+            drop_last_view() = default;
+            drop_last_view(Rng rng, difference_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+                return *sentinel;
+            }
+            template<typename CRng = Rng const,
+                CONCEPT_REQUIRES_(RandomAccessRange<CRng>())>
+            iterator_t<CRng> begin() const
+            {
+                return ranges::begin(rng);
+            }
+            template<typename CRng = Rng const,
+                CONCEPT_REQUIRES_(RandomAccessRange<CRng>())>
+            sentinel_t<CRng> end() const
+            {
+                return ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+            }
+
+            CONCEPT_REQUIRES(SizedRange<Rng>())
+            size_t size()
+            {
+                return get_size(rng);
+            }
+            CONCEPT_REQUIRES(SizedRange<Rng const>())
+            size_t size() const
+            {
+                return get_size(rng);
+            }
+
+            Rng & base()
+            {
+                return rng;
+            }
+            Rng const & base() const
+            {
+                return rng;
+            }
+        };
+
+        template<typename Rng>
+        struct drop_last_view<Rng, false>
+          : view_adaptor<
+              drop_last_view<Rng, false>,
+              Rng,
+              is_finite<Rng>::value ? finite : range_cardinality<Rng>::value   // finite at best (but unknown is expected)
+          >
+        {
+            CONCEPT_ASSERT(ranges::ForwardView<Rng>());
+        private:
+            friend range_access;
+
+            using difference_t = range_difference_type_t<Rng>;
+            using size_t = range_size_type_t<Rng>;
+            difference_t n;
+
+            template<bool IsConst>
+            struct adaptor : adaptor_base
+            {
+                using CRng = meta::const_if_c<IsConst, Rng>;
+                iterator_t<CRng> probe;
+
+                adaptor() = default;
+
+                template<bool OtherConst,
+                    CONCEPT_REQUIRES_(IsConst && !OtherConst)>
+                adaptor(adaptor<OtherConst> other)
+                    : probe(other.probe)
+                {}
+
+                adaptor(CRng &rng, difference_t n)
+                {
+                    probe = ranges::next(ranges::begin(rng), n, ranges::end(rng));
```

Comment:
This makes `begin()` O(N) always. The probe should be stored in a `non_propagating_cache` in `drop_last_view` and initialized on first access.

Reply:
Good point. Though only `probe_begin` should be stored in view. `probe` should still live in iterator (but now inited from `probe_begin`), since it need to be incremented on each `next()` step. Otherwise, each ++it will increase view's probe. 

And now view becomes mut only.